com.unity.shadergraph/Editor/Data/Nodes/AbstractMaterialNode.cs

```diff
@@ -761,16 +761,36 @@ public MaterialSlot AddSlot(MaterialSlot slot, bool findPrevInstance = true)
                 return foundSlot;
 
             // Try to keep the existing instance to avoid unnecessary changes to file
-            if (findPrevInstance && foundSlot != null && slot.GetType() == foundSlot.GetType())
+            if (attemptToModifyExistingInstance && foundSlot != null && slot.GetType() == foundSlot.GetType())
             {
                 foundSlot.displayName = slot.RawDisplayName();
                 foundSlot.CopyDefaultValue(slot);
                 return foundSlot;
             }
 
             // this will remove any old slots and then add the new one
-            m_Slots.RemoveAll(x => x.value.id == slot.id);
-            m_Slots.Add(slot);
+            // we keep the existing ordering by trying to replace the first match
+            bool insertedSlot = false;
+            for (int x = 0; x < m_Slots.Count; x++)
+            {
+                var existingSlot = m_Slots[x];
+                if (existingSlot.value.id == slot.id)
+                {
+                    if (!insertedSlot)
+                    {
+                        m_Slots[x] = slot;
+                        insertedSlot = true;
+                    }
+                    else
+                    {
+                        // this is a duplicate slot id... remove it
+                        m_Slots.RemoveAt(x);
+                        x--;
+                    }
+                }
+            }
+            if (!insertedSlot)
+                m_Slots.Add(slot);
```

Comment:
This seems a lot cleaner and C#-ish to me
```suggestion
            var foundIndex = m_Slots.FindIndex(x => x.value.id == slot.id);
            if(foundIndex < 0)
            {
                foundIndex = m_Slots.Count;
            } else {
                m_Slots.RemoveAll(x => x.value.id == slot.id);
            }
            m_Slots.Insert(foundIndex, slot);
            
```

Reply:
Not sure if it's cleaner, just more C# style, and slower..  :)

Reply:
Hmm..  if there was a RemoveAllFromRange() it would be expressable in C# in the near optimal form...

Reply:
How about this?
```

            int firstIndex = m_Slots.FindIndex(s => s.value.id == slot.id);
            if (firstIndex >= 0)
            {
                m_Slots[firstIndex] = slot;
                m_Slots.RemoveAllFromRange(s => s.value.id == slot.id, firstIndex + 1, m_Slots.Count - (firstIndex + 1));
            }
            else
                m_Slots.Add(slot);


// with RemoveAllFromRange function:

        public static int RemoveAllFromRange<T>(this List<T> list, Predicate<T> match, int startIndex, int count)
        {
            // TODO: to match behavior of RemoveRange we should throw exceptions when startIndex/Count are bad:

            int readIndex = startIndex;
            int writeIndex = startIndex;
            while (readIndex < list.Count)
            {
                T element = list[readIndex];
                readIndex++;

                bool remove = match(element);
                if (!remove)
                {
                    list[writeIndex] = element;
                    writeIndex++;
                }
            }

            // once we're done, we can remove the entries at the end
            int numberRemoved = list.Count - writeIndex;
            if (numberRemoved > 0)
            {
                list.RemoveRange(writeIndex, numberRemoved);
            }

            return numberRemoved;
        }




```

Reply:
Makes sense, let's put that in a ListUtilities file and ship it.